compiler/src/iree/compiler/Dialect/VMVX/Transforms/ResolveBufferDescriptors.cpp

```diff
@@ -17,20 +20,181 @@
 namespace mlir::iree_compiler::IREE::VMVX {
 
 namespace {
+/// Helper struct to return the offset, sizes and strides
+/// of a `source` of a `memref.extract_strided_metadata` op.
+struct DescriptorInfo {
+  OpFoldResult offset;
+  SmallVector<OpFoldResult> sizes;
+  SmallVector<OpFoldResult> strides;
+};
+}  // namespace
 
-struct FromMemRefSubView : public OpRewritePattern<GetBufferDescriptorOp> {
-  using OpRewritePattern::OpRewritePattern;
-  LogicalResult matchAndRewrite(GetBufferDescriptorOp op,
+/// Returns an AffineMap for an add or a mul.
+static AffineMap getAddMap(MLIRContext *context) {
+  AffineExpr s0, s1;
+  bindSymbols(context, s0, s1);
+  return AffineMap::get(0, 2, s0 + s1);
+}
+static AffineMap getMulMap(MLIRContext *context) {
+  AffineExpr s0, s1;
+  bindSymbols(context, s0, s1);
+  return AffineMap::get(0, 2, s0 * s1);
+}
+
+static FailureOr<DescriptorInfo> resolveBufferDescriptorForSubview(
+    memref::SubViewOp subview, RewriterBase &rewriter, Location loc,
+    Value sourceOffset, ValueRange sourceSizes, ValueRange sourceStrides) {
+  DescriptorInfo resultDescriptor;
+
+  // For sizes, we just use the new ones.
+  resultDescriptor.sizes = subview.getMixedSizes();
+
+  // Apply stride multipliers.
+  AffineMap mulMap = getMulMap(rewriter.getContext());
+  for (auto [index, stride] : llvm::enumerate(subview.getMixedStrides())) {
+    OpFoldResult currentStride = makeComposedFoldedAffineApply(
+        rewriter, loc, mulMap, {sourceStrides[index], stride});
+    resultDescriptor.strides.push_back(currentStride);
+  }
+
+  // Offsets.
+  resultDescriptor.offset = sourceOffset;
+  AffineMap addMap = getAddMap(rewriter.getContext());
+  for (auto [index, offset] : llvm::enumerate(subview.getMixedOffsets())) {
+    OpFoldResult physicalOffset = makeComposedFoldedAffineApply(
+        rewriter, loc, mulMap, {offset, resultDescriptor.strides[index]});
+    resultDescriptor.offset = makeComposedFoldedAffineApply(
+        rewriter, loc, addMap, {resultDescriptor.offset, physicalOffset});
+  }
+  return resultDescriptor;
+}
+
+/// Returns the strides based on the sizes assuming that the `memref`
+/// has default layout, i.e. it is not a result of a subview.
+static SmallVector<OpFoldResult> getStridesFromSizes(
+    RewriterBase &rewriter, Location loc, ArrayRef<OpFoldResult> sizes) {
+  if (sizes.size() == 0) {
+    return {};
+  }
+  SmallVector<OpFoldResult> strides(sizes.size());
+  strides.back() = rewriter.getIndexAttr(1);
+  AffineMap mulMap = getMulMap(rewriter.getContext());
+  for (int i = sizes.size() - 2; i >= 0; --i) {
```

Comment:
What happens if sizes.size() == 1?

Reply:
I think if sizes.size() == 1, i starts at -1 and loop doesnt execute. I think I copied it over from what was there. Ill add an early exit for size == 1